api/src/controllers/users.js

```diff
@@ -124,154 +125,146 @@ function UsersControllerFactory (auth, User, Invite, log, ledger, socket, config
+          // Otherwise an attacker could take over a ledger account
+          // for which no ILP kit account exits
+          const exists = yield co(ledger.existsAccount.bind(ledger, userObj))
+          if (!exists) {
+            yield co(ledger.createAccount.bind(ledger, userObj))
+          } else {
+            throw new Error(`Username ${userObj.username} already exists on the ledger` +
+              ', but not in the ILP kit.')
+          }
+        } catch (e) {
+          log.error(e)
+          throw new UsernameTakenError('Ledger rejected username')
+        }
+      })).then(co.wrap(function * (result) {
+        // transaction was commited
+        yield co(UsersController._onboardUser.bind(this, invite, dbUser))
+      }).bind(this)).catch(function (e) {
+        // transaction was rolled back
+        log.debug(e)
+        throw e
+      })
+    }
 
+    static * _handleInvite (invite, username) {
       try {
-        dbUser = yield dbUser.save()
-        dbUser = User.fromDatabaseModel(dbUser)
-
-        // Invite codes can only be used once
         if (invite) {
-          // Admin account funding the new account
-          const source = yield User.findOne({
-            where: {
-              username: config.data.getIn(['ledger', 'admin', 'user'])
-            }
-          })
-
-          // Send the invite money
-          yield pay.pay({
-            source: source.getDataExternal(),
-            destination: dbUser.username,
-            sourceAmount: invite.amount,
-            destinationAmount: invite.amount,
-            message: 'Claimed invite code: ' + invite.code
-          })
-
-          invite.user_id = dbUser.id
-          invite.claimed = true
-
-          yield invite.save()
+          if (invite.amount) {
+            // Admin account funding the new account
+            const source = yield User.findOne({
+              where: {
+                username: config.data.getIn(['ledger', 'admin', 'user'])
+              }
+            })
+
+            // Send the invite money
+            yield pay.pay({
+              source: source.getDataExternal(),
+              destination: username,
+              sourceAmount: invite.amount,
+              destinationAmount: invite.amount,
+              message: 'Claimed invite code: ' + invite.code
+            })
+            invite.claimed = true
+            invite.save()
+          }
         }
-
-      // TODO:SECURITY account should be funded at this point
       } catch (e) {
+        // TODO User did not receive his invite money
+        log.error(e)
         throw new ServerError()
       }
+    }
+
+    static * _onboardUser (invite, dbUser) {
+      if (invite) {
+        yield co(UsersController._handleInvite(invite, dbUser.username))
+      }
 
       // Fund the newly created account
-      yield UsersController.reload(dbUser)
+      yield co(UsersController.reload(dbUser))
```

Comment:
By default, sequelize uses bluebird as promise library and with bluebird it is a little tricky to yield subroutines. Bluebird's coroutine mechanism doesn't permit yielding generator instances directly. Since this is a little easier with `co`, I am using that instead.